### ISSUE TITLE
Unable to block innerHTML usage with DOM.setHTMLPolicy

### ISSUE THREAD

**Issue by shhnjk** [microsoft/fast#4118]
**Describe the bug; what happened?**
Following describes how one can sanitize `innerHTML` usage.
https://github.com/microsoft/fast/blob/2e730885a485febfbc2b993040f039ef4b85b4ec/packages/web-components/fast-element/docs/guide/declaring-templates.md#L144-L169

Since we didn't wanted our developers to use `innerHTML`, we've created following policy.
```
import { DOM } from '@microsoft/fast-element';
const myPolicy = trustedTypes.createPolicy('my-policy', {
  createHTML(html) {
    console.log(html);
    throw('You should never use innerHTML!!!');
  }
});
DOM.setHTMLPolicy(myPolicy);
```

However, this break FAST component itself.
Above `console.log` gives us following code (examples), which we don't think we ever produced using `innerHTML` usage in our code base:
```
    <template>
        fast-vkj8bh{0}fast-vkj8bh
        fast-vkj8bh{1}fast-vkj8bh
    </template>
```
or
```
   <template
        role="menu"
        @keydown="fast-vkj8bh{0}fast-vkj8bh"
        @focusout="fast-vkj8bh{1}fast-vkj8bh"
    >
        <slot fast-slotted="fast-vkj8bh{2}fast-vkj8bh"></slot>
    </template>
```

And we are suspecting that these are produced by FAST component itself.
https://github.com/microsoft/fast/blob/c93bc26c1fa438d86d049e6e0d8a09c65fda5781/packages/web-components/fast-element/src/template.ts#L89

How are we supposed to block/sanitize `innerHTML` usage as described in `declaring-templates.md` when FAST component itself uses `DOM.createHTML` which isn't aligned with our needs?

**Comment by shhnjk** [comment]
CC: @EisenbergEffect because this was added in https://github.com/microsoft/fast/issues/2850

**Comment by EisenbergEffect** [comment]
The purpose of this API is not to block the use of innerHTML but rather to ensure that any html that is rendered is properly sanitized. We could split this into two API potentially...one that is used by FAST and one that is specifically used for innerHTML scenarios. Thoughts?

**Comment by shhnjk** [comment]
Yes, splitting makes sense.
But is there any guarantee that anything generated by FAST won't contain user controllable data which would result in XSS? If there is nothing like a [Script Gadgets](https://github.com/google/security-research-pocs/tree/master/script-gadgets) in FAST, then the split should solve the issue 😊

**Comment by EisenbergEffect** [comment]
Thanks for the feedback. I'm adding this as a feature request and will put it pretty high up in our backlog.

**Comment by shhnjk** [comment]
Hi, after thinking about this bit more, I think the only Trusted Types policy required is for internal usage. The other possibility where developer might pass string to HTML sinks, you would just need to make sure that you are assigning that value as-is.

This way, if a developer passes a `TrustedHTML`, there won't be any violation. But if they pass string, there will be a violation. This is how React handles `dangerouslySetInnerHTML` for Trusted Typed environment as well.

Also, it'd be great if there's any update on this :)

**Comment by EisenbergEffect** [comment]
@shhnjk I want to make sure I understand you correctly before making the change. Are you saying that we should adjust this so that only `innerHTML` is handled with the policy and remove the usage during standard template creation?

**Comment by shhnjk** [comment]
Whenever Fast takes input from Web app to sinks which has Trusted Types enforcement, you will basically pass value as-is.

For example, we don't do the following:
```
 import { DOM } from '@microsoft/fast-element'; 
  
 let myPolicy = trustedTypes.createPolicy('my-policy', { 
   createHTML(html) { 
     // TODO: invoke a sanitization library on the html before returning it 
     return html; 
   } 
 }); 
  
 DOM.setHTMLPolicy(myPolicy); 
```

But we do the following instead:
```
 import { DOM } from '@microsoft/fast-element'; 
  
 const myPolicy = trustedTypes.createPolicy('my-policy', { 
   createHTML(html) { 
     // TODO: invoke a sanitization library on the html before returning it 
     return html; 
   } 
 }); 
 ...

 <div :innerHTML="${myPolicy.createHTML(someDangerousHTMLContent)}"></div> 
```

I'm not sure if this is technically possible, but if it is, that seems like a better option.

**Comment by EisenbergEffect** [comment]
It's definitely possible from a technical perspective to allow this. It puts more responsibility back in the hands of the developer, but also more flexibility. I don't know that there's a way to ensure that the innerHTML property is always set through a policy though, which means we couldn't protect developers from making dangerous mistakes. Due to security policy, we might not be able to enable that.

We could remove the use of the policy for standard templates, and instead only use it for innerHTML, which I believe would address the original issue. In a future major version bump, we could also make the default policy throw, so that developers must provide a policy if they use innerHTML.

Thoughts? /cc @nicholasrice

**Comment by EisenbergEffect** [comment]
After some further review, I think we need to:

* Remove the use of the policy for normal template scenarios. This probably doesn't make sense anyway.
* Use the policy only for innerHTML binding.
* Throw an exception in the default policy so that devs must apply their own sanitizer when binding innerHTML
* Update the documentation to reflect this change

If folks need separate policies, could they provide a noop policy for the default policy and then still manually apply their own policy as part of the binding itself, as described above?

Thoughts?

**Comment by shhnjk** [comment]
> * Remove the use of the policy for normal template scenarios. This probably doesn't make sense anyway.

Are you referring to the following? This seems like an Fast's internal usage of `innerHTML`. If so, the safeness of that `inner
HTML` should be guaranteed by Fast (because Fast is the implementer of that `innerHTML`). Thus, it makes sense for Fast to create a policy for that (and use sanitizer if necessary).
https://github.com/microsoft/fast/blob/c93bc26c1fa438d86d049e6e0d8a09c65fda5781/packages/web-components/fast-element/src/template.ts#L89

> * Use the policy only for innerHTML binding.

The only sad part for this is that, even if a developer doesn't use `innerHTML` binding, they have to bless the Trusted Types Policy because it'll be created. If there is a way to create this Trusted Types Policy only when `innerHTML` is used by developers, that would be great.

**Comment by EisenbergEffect** [comment]
I think we may be speaking past each other a bit. Let me see if I can say it another way. I think there are a few options for us:

### Option 1

* We have TWO policies.
* The first policy is used whenever a template is created and will simply pass through.
* The second policy is used for innerHTML and will throw by default.
* Devs can supply their own policy for either case but must supply one if binding innerHTML.

### Option 2

* We have ONE policy.
* We use the existing policy only for innerHTML and do not invoke it when creating normal templates.
* We make the default policy throw by default.
* Devs must supply their own policy if binding innerHTML.

### Option 3

* We have NO policies.
* We write documentation that instructs people to use trusted types when binding to innerHTML
* Maybe we can check the value before assigning to innerHTML and throw if it's not an instance of TrustedHTML?

### Option 4

* We have ONE policy.
* It's the same policy we have today but we don't use if for innerHTML anymore. We only use it for template construction.
* We write documentation that instructs people to use trusted types when binding to innerHTML
* Maybe we can check the value of innerHTML and throw if it's not an instance of TrustedHTML?

Any other options? Which one do we think is best? I'm leaning towards option 4 I think. Checking for a TrustedType instance will be a bit difficult if the browser doesn't yet supported trusted types. So, some research/experimentation is likely needed here.

**Comment by shhnjk** [comment]
Agreed, I also think we should go for option 4 :)

**Comment by shhnjk** [comment]
Any ETA on this?

**Comment by EisenbergEffect** [comment]
@shhnjk Thank you for bringing this back up and for your patience. We had a number of major projects to work on which caused this to drop lower down on our priority list, including the new design tokens systems, new Fluent UI refresh, and some critical perf work. We've got a little bit more to do on that over the next week or so and then it should be behind us and a good time for us to look at picking this back up. I've got an rfc out to make some improvements to the binding system, and I'm thinking that maybe this work should be done as part of that. That work is a minor breaking change, not likely to really affect folks, and this seems to fall in a similar category. Here's the rfc for the other binding changes: https://github.com/microsoft/fast/issues/5083 Let me know what you think.

**Comment by shhnjk** [comment]
Sounds good to me 😊

**Comment by EisenbergEffect** [comment]
Assigning this to the FASTElement 2.0 milestone, since it would constitute a breaking change. We're going to try to do this together with a few other breaking changes so we don't have to churn too much. I don't have a timeline on that, but hopefully we can start working on this soon.

**Comment by shhnjk** [comment]
Any update on the timeline?

**Comment by EisenbergEffect** [comment]
It's currently slated for FASTElement 2.0, which is in progress now. I don't have an ETA on that just yet.

**Comment by EisenbergEffect** [comment]
Option 4 is now implemented in the fast-element 2.0 branch. Please have a look and see if that will meet your need.
